### ISSUE TITLE
Store JSX in Store.

### ISSUE THREAD

**Issue by Escapado** [solkimicreb/react-easy-state#28]
Hey,

first of all, thanks for this great library. I started tinkering with it a bit and I have encountered an interesting error.

When I try to set jsx in an object like so:
`let myObj = {test: <span>Test</span>};`
then I can interpolate this object in my views just fine.

But when I wrap myObj in a store and try to call it, I get the following error:
```
TypeError: 'get' on proxy: property '_store' is a read-only and non-configurable data property on the proxy target but the proxy did not return its actual value (expected '#<Object>' but got '[object Object]')
```

This error occurs in react in 
```
  1132 |   }
  1133 | } else if (isValidElement(node)) {
  1134 |   // This element was passed in a valid location.
> 1135 |   if (node._store) {
  1136 |     node._store.validated = true;
  1137 |   }
  1138 | } else if (node) {
```

So I guess I can't store JSX in a Store. Is there any way around this?

**Comment by solkimicreb** [comment]
Thanks for the issue, I will fix this tomorrow and release a new patch version ☺️ I will also give you some explanation about the cause tomorrow. Going to sleep now.

**Comment by solkimicreb** [comment]
Do you store JSX in the store, because you expect it to be 'reactive' or is it a just a convenient place to store it (because it is together with the rest of the data)? Generally I would suggest you to do not store JSX in stores (just like you would not store it in React or Redux state).

The error is thrown because React elements have a read-only `_store` prop. You use the stored JSX in your render so the created React elements are [Proxy wrapped by Easy State](https://github.com/nx-js/observer-util/blob/master/src/handlers.js#L22) to make them reactive. This violates a Proxy get trap invariant (a Proxy returns another value for a read-only property, which mimics a forbidden write operations and an error is thrown).

I can fix this with a line or two (by checking the property descriptors before the Proxy wrapping), but I will have to check the perf impact before I can merge it.

If you just prefer to store everything in stores (and there are some parts which you use inside the render, but prefer to be not reactive), I will try to work out an elegant, general solution.

**Comment by Escapado** [comment]
Ah I see. I think it is mostly for convenience and I'd like it to be reactive. So my use case is this:

I try to write a little text adventure. I wrote a component which should output text as if someone was typing it, called TypeText. I have a character store and a quest store. Among other things, in the quest store I store quest segments like so:

```
const quests = store({
    segments: [
        {id: 1, text: <TypeText>Hello {character.name} </TypeText>}, 
        {id: 2, text: <TypeText>Where would you like to go?</TypeText>}, 
        ....}
     ], 
...})
```

I'd like to push them into a view based on some actions the player can trigger. Now of course I know I could just write plain text, and for example have a function that will search for {character.name} and call eval() and place that into a component when needed but I thought I could work around that.

**Comment by solkimicreb** [comment]
The fix is ready, I am perf profiling and studying your use case before I merge it in though 🙂 Hopefully  I can release it tomorrow.

**Comment by solkimicreb** [comment]
Could you put your code on GitHub for me to try it with the new version? It's also okay if you would rather keep it private 🙂

**Comment by Escapado** [comment]
Sure thing! I'll send it to you later this evening.

**Comment by solkimicreb** [comment]
Thanks for the access, this is super nice 🙂 I didn't expect the app to be so big and well structured, well done.

Why do you think that the current way is not reactive though? Just add `window.character = character;` in the Stores file. And try modifying `character.name` from the dev console after you made a few options. You will see that all existing questions update to reflect the new character name. Did you mean something else by 'reactive'?

It doesn't matter how the render function depends on a store prop, it will be re-rendered if a prop - used by it - changes. In this case the render function depends on a function (`text()`), which depends on a store prop. If the store prop changes the render will re-run.

On another note, maybe you could use more derived data. It's just personal taste but I would probably use a [prop getter](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/get), like
```
get text () {
  return character.name + "? I never heard that name before, where are you from?"
}
```

instead of `text: () => character.name + "? I never heard that name before, where are you from?"`. 

I am merging the fix for storing JSX, but I would still say to avoid it whenever possible. Keeping the store as simple as possible makes your app more predictable. By simplicity I mean not storing third party data in the store, that you have no control over - which is the hidden complexity of JSX in your case.

Thanks for the issue, it was really helpful. I will keep it open until the release of the fix 🙂

**Comment by Escapado** [comment]
Thanks for your feedback!!

You are right, it is reactive, I just didn't think it was! You suggestions are valuable and I'll try to stick to using prop getters and JSX-less stores!  It's cleaner after all and I can still do things in the views that I was trying to do and it's probably much cleaner!

By the way, you rock (and so does this library!).

**Comment by solkimicreb** [comment]
Thanks for the kind words 🙂 I released [v4.1.2](https://github.com/nx-js/observer-util/releases/tag/v4.1.2) of the underlying reactivity library, which fixes the issue with none writable, none configurable props.

I let semver and transient deps do their jobs, so Easy State doesn't get a new release. You will need to update your package-lock, if you use any to receive the fix from the underlying lib. From now on you can store JSX in stores 🙂 Let me know, if you still have issues with this in the future.

Thanks for the issue, you found a sneaky bug 🙂
